### DISCUSSION TITLE
Server certificate from computer store

### DISCUSSION THREAD

**Discussion by IgorMilavec** [ravendb/ravendb#12858] [General]
My computer has an autoenrolled certificate in its certificate store (template DomainComputer). Using this certificate is IMO the most secure option, as I don't have to have PFX+password lying around. Unfortunately, it seems to me that RavenDB is unable to use a certificate directly from the computer store? Did I miss some configuration option?

**Comment by ayende** [comment]
You just need to load it into a `X509Certificate2` instance. 
See here for example, RavenDB can make use of it then:

https://stackoverflow.com/questions/6304773/how-to-get-x509certificate-from-certificate-store-and-generate-xml-signature-dat

**Reply by IgorMilavec** [reply]
Thank you for your answer. Unfortunately I failed to mention that I want to use the certificate from the computer's certificate store on the RavenDB server. I want RavenDB server to use such a certificate for TLS on its ServerUrl. As far as I can see, the server requires PFX (so the whole keypair)? If it is possible, can you please provide a sample settings.json stanza for this?

**Comment by ayende** [comment]
Yes, server side behavior like that is supported, see the documentation here:

https://ravendb.net/docs/article-page/5.2/csharp/server/security/authentication/certificate-configuration#authentication-manual-certificate-configuration

**Reply by IgorMilavec** [reply]
OK, but all the options require PFX... I do not have access to the PFX and cannot export a PFX from the computer store, because the private key is not exportable.

**Reply by ayende** [reply]
Did you see this example?
![image](https://user-images.githubusercontent.com/116915/137761520-952faa31-01de-4660-a2f3-24f5ff9a93e8.png)

This is a power shell script that will pull the certificate from the user's certificate store. Not PFX file in sight.

**Reply by ayende** [reply]
Oh, I see the issue. We do need this exportable, otherwise it won't work.

**Reply by IgorMilavec** [reply]
Hm, I was afraid this would be the answer. :(
Just out of my curiosity, why do you actually need the private key? I had a quick look at the source code and could not find where it is used.

**Reply by ayende** [reply]
We use TLS (SSL) for authentication, which means that we need to send the full certificate to the `SslStream` to be able to authenticate as a server.

See here for full discussion: https://ayende.com/blog/178977/ravendb-4-0-securing-the-keys-to-the-kingdom

**Comment by IgorMilavec** [comment]
I have read the blog post, but I do not see why the private key needs to be exportable. SslStream can use certificate directly from the computer store,  provided that the user has at least read permissions on the private key:
```c#
X509Store store = new(StoreName.My, StoreLocation.LocalMachine);
store.Open(OpenFlags.ReadOnly);
X509Certificate2 serverCertificate = store.Certificates
    .Find(X509FindType.FindBySubjectName, "raven.somewhere.local", validOnly: true)
    .Find(X509FindType.FindByApplicationPolicy, "1.3.6.1.5.5.7.3.1", false) // Server Authentication
    .OfType<X509Certificate2>()
    .Where(c => c.HasPrivateKey)
    .OrderByDescending(c => c.NotBefore)
    .FirstOrDefault();
if (serverCertificate is null)
{
    throw new Exception("A suitable certificate was not found.");
}

TcpListener listener = new(IPAddress.Any, 8443);
listener.Start();
TcpClient client = listener.AcceptTcpClient();
SslStream sslStream = new(client.GetStream(), leaveInnerStreamOpen: false);
sslStream.AuthenticateAsServer(serverCertificate, clientCertificateRequired: true, checkCertificateRevocation: false);


```

**Reply by ayende** [reply]
We have chosen to simplify our system in this regard. The sheer number of ways that you can store certificates is _staggering_.
As such, we accept the certificates in one of two ways:

* As a file on disk
* As the output of a script that does its own logic.

We have chosen that over code such as yours because in terms of support and troubleshooting, this make it a lot easier.
For example, in the case of your code, imagine just how many configuration options we'll need to expose this.
Or consider the number of silent failures that may happen here, which would be really hard to trouble shoot.

**Reply by IgorMilavec** [reply]
I cannot argue against "support and troubleshooting" as this is a business decision.
I would just like to point out that this decision makes life really hard for ops wanting to deploy RavenDB on-prem in a secure way:
 - file on disk is a non-starter for security reasons
 - having exportable cert in computer or user store is questionable in some environments
 - there is no Azure Key Vault on-prem (or you need to have the secret on the disk, which again is not secure)

Basically I see no usable way of having secure identity based access to the private key, apart from deploying some on-prem AD integrated key vault.
Windows certificate store supports this with very low friction. The only configuration option in my sample is the FQDN of the server (which is already configured in RavenDB anyway). Or you could go with requiring certificate thumbprint in the config. Any ops person supporting any Kestrel based app will have experience with this. And the cost is one optional configuration setting.
Just my 2 cents...

**Reply by ayende** [reply]
This won't work for running on Linux, which is the majority of our users.
This will fail when you have multiple matching certs with different ranges, you'll need a way to select which one you want.
It will also fail when you need to do automatic refresh of the certificate.

There is quite a set of features around certificate management in RavenDB, including things like automatically getting them from Let's Encrypt, updating from vaults / HSM, etc. 

In terms of the balance of complexity / features, we have selected to require that the user will provide us with the certificate itself in any way they want, and we'll take it from there.
